plugins/source.go

```diff
@@ -148,39 +144,40 @@ func (p *SourcePlugin) Sync(ctx context.Context, spec specs.Source, res chan<- *
 		if table.Multiplex != nil {
 			clients = table.Multiplex(c)
 		}
-		// we call this here because we dont know when the following goroutine will be called and we do want an order
-		// of table by table
-		// totalClients := len(clients)
-		// newN, err := helpers.TryAcquireMax(ctx, goroutinesSem, int64(totalClients))
-		// if err != nil {
-		// 	p.logger.Error().Err(err).Msg("failed to TryAcquireMax semaphore. exiting")
-		// 	break
-		// }
-		// goroutinesSem.TryAcquire()
-		w.Add(1)
-		go func() {
-			defer w.Done()
-			wg := sync.WaitGroup{}
-			p.logger.Info().Str("table", table.Name).Msg("fetch start")
-			tableStartTime := time.Now()
-			totalTableResources := 0
-			for _, client := range clients {
-				client := client
-
-				// i := i
-				wg.Add(1)
-				go func() {
-					defer wg.Done()
-					// defer goroutinesSem.Release(1)
-					totalTableResources += table.Resolve(ctx, client, fetchTime, nil, res)
-				}()
-			}
-			wg.Wait()
-			totalResources += totalTableResources
-			p.logger.Info().Str("table", table.Name).Int("total_resources", totalTableResources).TimeDiff("duration", time.Now(), tableStartTime).Msg("fetch table finished")
-		}()
-	}
-	w.Wait()
+
+		for _, client := range clients {
+			client := client
+			tableJobs = append(tableJobs, workerJob{
+				ctx:      ctx,
+				logger:   &p.logger,
+				client:   client,
+				table:    table,
+				syncTime: syncTime,
+				res:      res,
+			})
+		}
+	}
+
+	jobsCount := len(tableJobs)
+	jobs := make(chan workerJob, jobsCount)
+	results := make(chan workerResult, jobsCount)
+
+	for w := uint64(1); w <= concurrency; w++ {
+		go worker(jobs, results)
+	}
+
+	for i := 0; i < jobsCount; i++ {
+		jobs <- tableJobs[i]
+	}
+	close(jobs)
+
+	totalResources := 0
+	for i := 0; i < jobsCount; i++ {
+		result := <-results
```

Comment:
I would make the channels unbuffered and collect in a separate goroutine, to save on memory.

Reply:
I'll make this change once @yevgenypats approves of the new approach